realsense2_camera/src/base_realsense_node.cpp

```diff
@@ -1048,6 +1074,152 @@ void BaseRealSenseNode::publishStaticTransforms()
+{
+	static float num_valid;
+	static bool unit_vector_mature;
+	static unsigned int height, width, sequence;
+	static float current_unit_vectors[FRAME_PRIME_HEIGHT][FRAME_PRIME_WIDTH * 3]= {0};
+
+	try
+	{
+		if (!is_frame_arrived.at(DEPTH))
+		{
+			ROS_DEBUG("Brain: Skipping publish unit vector topic! Depth frame didn't arrive.");
+			return;
+		}
+	}
+	catch (std::out_of_range)
+	{
+		ROS_DEBUG("Brain: Skipping publish unit vector topic! Depth frame didn't configure.");
+		return;
+	}
+
+	auto image_depth16 = reinterpret_cast<const uint16_t*>(_image[DEPTH].data);
+	auto depth_intrinsics = _stream_intrinsics[DEPTH];
+
+	height = depth_intrinsics.height;
+	width = depth_intrinsics.width;
+
+	if ((detailed_logging) && (max_log-- > 0))
+		ROS_INFO("Brain: depth_intrinsics: width:%d; height:%d; _depth_scale_meters:%.6f.\n",
+					width, height, _depth_scale_meters);
+
+	if (!unit_vector_mature) {
+		float scaled_depth, num_bad_pixel = 0;
+
+		for (int y = 0; y < depth_intrinsics.height; ++y)
+		{
+			for (int x = 0; x < depth_intrinsics.width; ++x)
+			{
+				scaled_depth = static_cast<float>(*image_depth16) * _depth_scale_meters;
+
+				++image_depth16;
+
+				if (fabs(scaled_depth) < 0.0001) {
+					num_bad_pixel++;
+					continue;
+				}
+
+				if (valid_pixels[y][x]) {
+					continue;
+				}
+
+				unit_vectors[y][x * 3] = (x - depth_intrinsics.ppx) / (depth_intrinsics.fx * scaled_depth);
+				unit_vectors[y][x * 3 + 1] = (y - depth_intrinsics.ppy) / (depth_intrinsics.fy * scaled_depth);
+				unit_vectors[y][x * 3 + 2] = 1;
+
+				valid_pixels[y][x] = true;
+				num_valid++;
+			}
+		}
+
+		/* check the % of pixels with valid depth value in this frame */
+		if ((detailed_logging) && (max_log > 0))
+			ROS_INFO("Brain: %% of pixels in this frame with valid depth values is:%f %%.\n",
+						1 - num_bad_pixel / (depth_intrinsics.height * depth_intrinsics.width));
+
+		/* check the % of valid vectors in unit vectors array */
+		float rate = num_valid / (depth_intrinsics.height * depth_intrinsics.width);
+		if (rate >= mature_threshold) {
+			unit_vector_mature = true;
+			ROS_INFO("Brain: %% of pixels with valid depth values is:%f %% which is above the threshold %f %% .\n",
+						rate, mature_threshold);
+
+			ROS_INFO("Brain: prepare the data to be formatted to image message.\n");
+
+			for (unsigned int i = 0; i < height; i++)
+				for (unsigned int j = 0; j < width * 3; j++)
+					current_unit_vectors[i][j] = unit_vectors[i][j];
+
+			if (enable_unit_test)
+				output_unit_vector(height, width, unit_vectors);
+		} else {
+			if ((detailed_logging) && (max_log > 0))
+				ROS_INFO("Brain: %% of pixels with valid depth values is:%f %% which is below the threshold %f %% .\n",
+							rate, mature_threshold);
+		}
+	}
+
+	if (unit_vector_mature || !publish_high_quality) {
+		std::shared_ptr<cv::Mat> image;
+		float any_size_unit_vectors[height][width * 3]= {0};
+		void *data;
```

Comment:
I guess it depends how `data` is getting forwarded from float[][] to cv::Mat to cv_bridge::CvImage, but is it safe to clean up after we're done using `data`? (i.e. `data = nullptr`)